components/provisioning/org.wso2.carbon.identity.provisioning/src/main/java/org/wso2/carbon/identity/provisioning/OutboundProvisioningManager.java

```diff
@@ -1105,4 +1118,54 @@ private boolean userHasProvisioningRoles(List<String> userRoles, List<String> pr
 
         return false;
     }
+
+    /**
+     * Check whether it is needed to wait for provisioning operation getting completed.
+     * Even in the non-blocking provisioning, In Self Registration flow and Ask Password flow it is necessary to wait
+     * till the user provisioning get completed. The reason for this is that the User provisioning identifier receives
+     * only after the successful user provisioning. So if we don't wait till the user provisioning get completed, if a
+     * user claim update happens in IS side, since we don't have a user provisioning identifier in the DB, we are trying
+     * to create the user again instead of updating the user. This is critical in Self registration and Ask password
+     * flows, since during the user creation process we are doing a claim update also for lock the user account.
+     *
+     * @param provisioningEntity Provisioning Entity object.
+     * @return true if it is needed to wait(block) till the provisioning operation get completed, false otherwise.
+     */
+    private boolean needToWaitForUserProvisioning(ProvisioningEntity provisioningEntity) {
+
+        if (!ProvisioningEntityType.USER.equals(provisioningEntity.getEntityType()) ||
+                !provisioningEntity.getOperation().equals(ProvisioningOperation.POST)) {
+            return false;
+        }
+
+        // Need to wait in the self registration flow.
+        Map<ClaimMapping, List<String>> attributes = provisioningEntity.getAttributes();
+        if (attributes != null) {
+            for (ClaimMapping claimMapping : attributes.keySet()) {
```

Comment:
it might be more efficient to use entrySet() instead of keySet() in both the places

Reply:
Addressed with https://github.com/wso2/carbon-identity-framework/pull/5430/commits/a2da93f0993b9f5f14eb818986757170cc3c861f